### ISSUE TITLE
L>M Upgrade: Overrides are not correctly migrated

### ISSUE THREAD

**Issue by BjoernT** [rcbops/rpc-openstack#1485]
###### Please provide the output to 'git describe --all' for  all deployments where you have confirmed it occurs.

r13.1.0rc1
13.3.4-1-ga489036

##### What type of issue is this? Please tick the appropriate box.
- [x] bug - existing functionality that does not work for you as documented
- [ ] feature - new functionality that you would like to see added

##### What is the expected behavior?

All overrides changed in Liberty should be correctly migrated to Mitaka style overrides

##### What is actually happening?

user_osa_variables_overrides.yml contained default settings for maas_auth_method, maas_scheme, hp_check rather then their previous state on liberty:

```
/etc/openstack_deploy.liberty/user_extras_variables.yml:maas_auth_method: token
/etc/openstack_deploy.liberty/user_extras_variables.yml:hp_check: true
/etc/openstack_deploy.liberty/user_extras_variables.yml:maas_scheme: https
```

Also we were using Keystone default settings for  keystone_token_driver, keystone_token_provider
to switch to fernet. After the upgrade, UUID token was setup.

I would much rather prefer to populate 

```
keystone_token_driver: memcache
keystone_token_provider: fernet
```

one fernet token provider is detected.
 
Furthermore the db and rpc settings got lost too:

```
neutron_api_workers:
/etc/openstack_deploy/user_osa_variables_defaults.yml:neutron_api_workers: 64
/etc/openstack_deploy/user_osa_variables_overrides.yml:neutron_api_workers: 64
/etc/openstack_deploy.liberty/user_variables.yml:neutron_api_workers: 128
```

##### What were you attempting to do when you encountered this error (give as much detail as possible)? 

Updated the values inside user_rpco_variables_overrides.yml to old values

##### How can we reproduce this issue?

Upgrade L to M

##### How can we verify your issue is addressed?

Old settings are correctly retained

##### If you are using or know of a workaround please describe the necessary steps?

N/A

##### What is your evaluation on the importance of this issue, and why?

Prio 1

----------------------------------------------------------------------------------------------------------------------------
###### For each pull request, update the branch line to reference it.
- [ ] master
- [ ] mitaka-13.1

##### Upgrade impact. Consider the impact on moving between major, minor and patch releases.
- [ ] This change does not impact upgrades
- [ ] This change addresses a potential upgrade impact
- [ ] The upgrade impact is being tracked by another issue.

The issue tracking the upgrade impact is:

##### Testing impact
- [ ] An issue/pull-request has been create or exists to track the testing gap, and is referenced below
- [ ] This change includes additional testing
- [ ] This change is tested by the existing setup

Additional testing links:

##### Release note
- [ ] This change includes a release note.
- [ ] This change does not require a release note.

##### Does your change require documentation?
- [ ] Yes, and I submitted an issue to [Docs issues](https://github.com/rackerlabs/docs-rpc/issues "Docs issues")
- [ ] Yes, and I contributed a documentation patch to [docs-rpc](https://github.com/rackerlabs/docs-rpc "docs-rpc")
- [ ] No documentation required for this change

Additional documentation links (please add links if you have ticked yes):

**Comment by alextricity25** [comment]
@BjoernT, it sounds like you might of ran the ``test-upgrade.sh`` script on this environment, and thus causing the variables to take the values of the new defaults. 
The ``test-upgrade.sh`` and ``migrate-yaml.py`` script use [a flag](https://github.com/rcbops/rpc-openstack/blob/master/scripts/test-upgrade.sh#L34), ``--for-testing-take-new-vars-only``, to automate the migration of variables for our gating jobs and dev purposes. This was by design so we could have an automated way for testing upgrades. However, this flag(nor the ``test-upgrade.sh`` script) shouldn't be used for production environments, and migrating variables in such environments is actually a fairly manual process. This is due to the fact that assumptions can't be made on the deployers behalf. If there are different default values in the newer release, do we assume those? Or keep the old values? These are decisions we felt would be better put in the deployer's hand.
 Documentation for this process is in a WIP PR [here](https://github.com/rackerlabs/docs-rpc/pull/620/files#diff-dd39230c5ac09e679cf15c660ba2de64R22). Please follow these instructions and let us know if you run into any problems!

**Comment by BjoernT** [comment]
@alextricity25 That would explain it. Good to know, I was told to use this script in absence of a finished upgrade script. Closing it out for now

**Comment by BjoernT** [comment]
Please reopen this issue is happening with without the option --for-testing-take-new-vars-only too. maas_auth_token is now entirely missing and only present inside OLD_OVERRIDES section

**Comment by alextricity25** [comment]
@BjoernT Would it possible to get a copy of your ``user_extras_variables.yml`` file? I would like to see if I can reproduce this.

**Comment by BjoernT** [comment]
yes I mail it to you, it's from a lab

**Comment by git-harry** [comment]
@BjoernT from your description this sounds like it is working as intended.

user_extras_variables.yml (file on Liberty deployment)
```
---
maas_auth_token: customer_token
```

user_rpco_variables_defaults.yml (Mitaka defaults file)
```
---
maas_auth_token: some_token
```

migrate-yaml.py script output (user_rpco_variables_overrides.yml)
```
---
NEW_DEFAULTS:
  maas_auth_token: some_token
OLD_OVERRIDES:
  maas_auth_token: customer_token
```

NEW_DEFAULTS is the value in the Mitaka defaults file user_rpco_variables_defaults.yml.
OLD_OVERRIDES is the value in the Liberty variables file user_extras_variables.yml.

The script outputs the values where the operator needs to make a decision as to which is the appropriate value to keep. In this instance, after editing the file ,user_rpco_variables_overrides.yml, would look like:

```
---
maas_auth_token: customer_token
```

/cc @prometheanfire @alextricity25

**Comment by alextricity25** [comment]
@BjoernT I never got it. Did you send it to miguel.cantu@rackspace.com?

**Comment by BjoernT** [comment]
Please leave it open until figured out, this is supposed to be automated to a point where this process happens inside the run-upgrade script. There is no period inside the script where we actually check the variables, the upgrade runs immediately after the conversion. 
Also If I update maas_auth_token before starting the run-upgrade.sh script (or test-upgrade with modification in this case) then I expect that the token is correctly added to the user_rpco_overrides.yaml file

**Comment by git-harry** [comment]
@BjoernT you misunderstand how these things are designed to work. I will cover this of in the meeting shortly.

**Comment by BjoernT** [comment]
@git-harry Well then there are not designed to the extend what I had discussed with  @prometheanfire the implementation is insufficient in this case

**Comment by prometheanfire** [comment]
The problem with doing this variable migration in a fully automated way is that we don't KNOW for sure if the old variable or the new variable should be taken.  Once the migration is done though we know we can statically always update the defaults files as we control them.

**Comment by git-harry** [comment]
There is still some confusion on the support side, set to undecided pending additional information from @rcbops/support-feedback

**Comment by alextricity25** [comment]
@BjoernT I ran the ``migrate-yaml.py`` script with the ``user_extras_variables.yml`` file you sent me, and it looks like it ran as intended.
Here is the command I ran:
```
./migrate-yaml.py --defaults ../rpcd/etc/openstack_deploy/user_rpco_variables_defaults.yml \
> --overrides user_extras_variables.yml.bak --output-file test_variables.yml
```

And the output(sensitive data stripped):
```
---
NEW_DEFAULTS:
  maas_auth_method: password
  maas_auth_token: some_token
  maas_scheme: http
  rackspace_cloud_api_key: SomeAPIKey
  rackspace_cloud_password: SomeUsersPassword
  rackspace_cloud_tenant_id: SomeTenantID
  rackspace_cloud_username: SomeUserName
OLD_OVERRIDES:
  maas_auth_method: token
  maas_auth_token: *********************
  maas_scheme: https
  rackspace_cloud_api_key: *********************
  rackspace_cloud_password: *********************
  rackspace_cloud_tenant_id: *********************
  rackspace_cloud_username: *********************
lb_name: *********************
maas_agent_token: *********************
maas_api_url: *********************
maas_fqdn_extension: *********************
```
It does look like ``maas_auth_method`` is both in ``NEW_DEFAULTS`` and ``OLD_OVERRIDES``. At this point it would be up to you to choose which values you want to keep. I imagine for your case the file would be changed to look like:
```
maas_auth_method: token
maas_auth_token: *********************
maas_scheme: https
rackspace_cloud_api_key: *********************
rackspace_cloud_password: *********************
rackspace_cloud_tenant_id: *********************
rackspace_cloud_username: *********************
lb_name: *********************
maas_agent_token: *********************
maas_api_url: *********************
maas_fqdn_extension: *********************
```
Notice how the ``NEW_DEFAULTS`` and ``OLD_OVERRIDES`` blocks have been removed and the variables have been properly indented.

I realize for these specific variables, it might not make a lot of sense why we are not automating the process. But just think about situations where the deployer has a different value in the liberty configurations than the what the mitaka default configurations have. In some situations they might want the new default value, in others they might want to keep the old override.

**Comment by BjoernT** [comment]
OK I will close this issue and manually listing the steps so we can verify the migrated overrides before executing the upgrade
